pkg/chains/storage/ca/ca.go

```diff
@@ -0,0 +1,350 @@
+	logger         *zap.SugaredLogger
+	tr             *v1beta1.TaskRun
+	client         pb.GrafeasV1Beta1Client
+	cfg            config.Config
+	occurrenceRefs []string // store occurrence IDs that are automatically generated during the time of creation
+}
+
+func NewStorageBackend(logger *zap.SugaredLogger, tr *v1beta1.TaskRun, cfg config.Config) (*Backend, error) {
+	// add TypeMeta information to taskrun objects
+	addTypeInformationToObject(tr)
+
+	// build connection through grpc
+	// implicit uses Application Default Credentials to authenticate.
+	// Requires `gcloud auth application-default login` to work locally
+	ctx := context.Background()
+	creds, err := oauth.NewApplicationDefault(ctx, "https://www.googleapis.com/auth/cloud-platform")
+	if err != nil {
+		return nil, err
+	}
+
+	conn, err := grpc.Dial("dns:///containeranalysis.googleapis.com",
+		grpc.WithTransportCredentials(credentials.NewTLS(&tls.Config{})),
+		grpc.WithDefaultCallOptions(grpc.PerRPCCredentials(creds)),
+	)
+	if err != nil {
+		return nil, err
+	}
+	// defer conn.Close()
+
+	// connection client
+	client := pb.NewGrafeasV1Beta1Client(conn)
+
+	// create backend instance
+	b := &Backend{
+		logger:         logger,
+		tr:             tr,
+		client:         client,
+		cfg:            cfg,
+		occurrenceRefs: []string{},
+	}
+
+	// create note
+	err = b.createNote()
+
+	if err != nil {
+		// ignore note already exisits error
+		logger.Info(err)
+	}
+
+	return b, nil
+}
+
+// StorePayload implements the storage.Backend interface.
+func (b *Backend) StorePayload(rawPayload []byte, signature string, opts config.StorageOpts) error {
+	b.logger.Infof("Trying to store payload on TaskRun %s/%s", b.tr.Namespace, b.tr.Name)
+
+	// We only support simplesigning for OCI images, and in-toto for taskrun.
+	if opts.PayloadFormat == formats.PayloadTypeTekton || opts.PayloadFormat == formats.PayloadTypeProvenance {
+		return errors.New("Container Analysis storage backend only supports for OCI images and in-toto attestations")
+	}
+
+	// step1: create occurrence request
+	occurrenceReq := b.createOccurrenceRequest(rawPayload, signature, opts)
+	// step2: create/store occurrence
+	occurrence, err := b.client.CreateOccurrence(context.Background(), occurrenceReq)
+	if err != nil {
+		return err
+	}
+
+	// store reference to the newly generated occurrence for retrieve purpose
+	b.occurrenceRefs = append(b.occurrenceRefs, occurrence.GetName())
+
+	b.logger.Infof("Successfully created an occurrence %s (Occurrence_ID is automatically generated) for Taskrun %s/%s", occurrence.GetName(), b.tr.Namespace, b.tr.Name)
+	return nil
+}
+
+// Retrieve payloads from container analysis and store it in a map
+func (b *Backend) RetrievePayloads(opts config.StorageOpts) (map[string]string, error) {
+	// initialize an empty map for result
+	result := make(map[string]string)
+
+	// get all occurrences created using this backend
+	occurrences, err := b.getOccurrences()
+	if err != nil {
+		return nil, err
+	}
```

Comment:
IIUC, this is relying on the `Backend.occurrences` field, which means that this will only work if the instances reading the occurrences is the same as the one that wrote it. I don't think we should make this assumption - it limits usage of this library (e.g. what if I want to retrieve the payloads in a CLI or something else that is only reading the results).

Reply:
That's true. I was only looking at the place where the retrieve functions are used [here](https://github.com/tektoncd/chains/blob/8b29855fca5574de429fb20234cf66aefe003119/pkg/chains/verifier.go#L57-L75) in verifier.go. In verifier, as a receiver function of the backend, `RetrievePayloads` retrieves payload for each backend instance of each `signableType`. So for taskrun artifact, it only reads the occurrence of taskrun artifact attestation. And for OCI artifact, it reads the occurrences of OCI images outputted from a taskrun.

> it limits usage of this library (e.g. what if I want to retrieve the payloads in a CLI or something else that is only reading the results).

Can you elaborate a bit more here?